### ISSUE TITLE
SDP support for JupyterHub

### ISSUE THREAD

**Issue by adwk67** [stackabletech/issues#191]
As a machine learning engineer, I want to test/asses the viability of using JupyterHub notebooks from within a kubernetes environment. This involves the following:

- how does [JuypterHub](https://jupyterhub.readthedocs.io/en/stable/) integrate with Kubernetes?
  - authentication/authorization
  - resource management
  - external resources
- ~is it possible to deploy JupyterHub via an existing Helm chart, or is an operator required?~ Yes, we use it in demos
- what are the possibilities for interacting with kubernetes-managed clusters directly from a notebook environment?

The result of this story will be:

- documentation outlining the findings
- a decision on how to proceed
- optionally (dependent on previous point): implementation steps

Links:
- https://github.com/stackabletech/issues/issues/653
- https://jupyter-docker-stacks.readthedocs.io/en/latest/using/selecting.html#jupyter-pyspark-notebook
- https://quay.io/repository/jupyter/pyspark-notebook?tab=tags

From the a [comment](https://github.com/stackabletech/decisions/issues/36#issuecomment-2535355541) in the [decision](https://github.com/stackabletech/decisions/issues/36):

Remarks / Questions / Priorities:

| Demo | Tutorial | Stage 1 tasks | To demonstrate | Effort | Remarks | Priority | 
|  ------------- |  ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| 🟢  | 🟢  | Custom images  | images that co-ordinate jupyterhub, SDP spark images and python | M | Proposal: we document how to build custom images on top of official spark ones, which do not have jar/python compatibility issues with the notebook images| 1 |
| | | | extend these images to offer a bundle of typical libraries (e.g. pandas, numpy, sklearn, keras)  | L | We suggest to deliver one "Basic Data Science Image" (details tbd, effort size M?! ) and extend them only on customer request (as individual agreement)  |  12 |
|  | | Notebook integration with source control | directly with terminal | S |  |  6 |
|  | | | gitsync (required to be in the image) | L | |  11 |
|  | | | vscode integration | L | |  7 |
|  | 🟢  | Authentication | native authentication | S | (will document only, incl. changes between 3.0.0 and 4.0.0)| 4 | 
| 🟢  | 🟢  | | keycloak authentication | M/L | (Only) Keycloak as implementation of GenericOAuthenticator | 5 |
| 🟢  | 🟢 | Profiles | image selection | S |   | 9 | 
| 🟢  | | | storage selection | S |  | 8 |
| 🟢  | 🟢 | | resource settings (requests and limits) | S |  | 10  |
|  | | | load external resources (via lifecycle hook) | M |  | 13 |
| 🔴 | 🔴 | Interaction with storage components | HDFS | M/S | (removed: we have this in the jupyterhub-pyspark demo)  | 3  |
| 🟢  | 🟢  | | S3 | M |  | 2 |

**Comment by adwk67** [comment]
A JupyterHub helm install creates the following manifests:

```
kind: NetworkPolicy - name: hub
kind: NetworkPolicy - name: proxy
kind: PodDisruptionBudget - name: user-placeholder
kind: ServiceAccount - name: hub
kind: Secret - name: hub
kind: ConfigMap - name: hub
kind: PersistentVolumeClaim - name: hub-db-dir
kind: Role - name: hub
kind: RoleBinding - name: hub
kind: Service - name: hub
kind: Service - name: proxy-api
kind: Service - name: proxy-public
kind: Deployment - name: hub
kind: Deployment - name: proxy
kind: StatefulSet - name: user-placeholder
```

**Comment by adwk67** [comment]
# Proposal

- **Must have**: a _suite of images_ that are consistent across python/jupyterhub/spark versions
- **Should have**: _Authentication with Keycloak_/OIDC
- **Could have**: 
  - _Authorization with OPA_
  - a jupyterhub-_operator_

These points are outlined below.

## A suite of images

See [this](https://github.com/stackabletech/issues/issues/653) issue for more information. 
If we are to offer any formal support for JupyterHub, it is essential that we can recommend images that work across this version matrix. It would not involve any new code. 

## Authentication with Keycloak

A WIP can be found in this [branch](https://github.com/stackabletech/demos/tree/feat/keycloak-jupyterhub/stacks/jupyter-keycloak).
Keycloak can be set up with a **`jupyterhub`** client/realm, much like is [done](https://github.com/stackabletech/demos/blob/feat/keycloak-jupyterhub/stacks/end-to-end-security/keycloak-realm-config.yaml#L227-L248) in the e2e-security demo.
See also https://oauthenticator.readthedocs.io/en/latest/tutorials/provider-specific-setup/providers/generic.html.
The JuypterHub config should reference **`GenericOAuthenticator`**:
```
hub:
    config:
      GenericOAuthenticator:
        client_id: jupyterhub
        client_secret: jupyterhubjupyterhub
        # TODO: How to define this here before Jupyterhub has been deployed?
        oauth_callback_url: http://172.19.0.3:32336/hub/oauth_callback
        # TODO: template the host:port for Keycloak
        authorize_url: https://172.19.0.3:31245/realms/jupyterhub/protocol/openid-connect/auth
        token_url: https:///172.19.0.3:31245/realms/jupyterhub/protocol/openid-connect/token
        userdata_url: https://172.19.0.3:31245/realms/jupyterhub/protocol/openid-connect/userinfo
        username_claim: preferred_username
        auth_state_groups_key: "oauth_user.groups"
        manage_groups: True
        tls_verify: False
        allowed_groups:
          - "Notebook Developers"
        scope:
          - openid
      JupyterHub:
        authenticator_class: generic-oauth
```

## Authorization with OPA

The [docs](https://z2jh.jupyter.org/en/latest/administrator/authentication.html#authentication-and-authorization) seem to tackle by lumping both custom Authentication and Authorization in a custom Authenticator class. 

### Authenticate/Authorize on login
Provide a custom Authenticator and issue a callout to the OPA server in the [authenticate](https://jupyterhub.readthedocs.io/en/stable/reference/api/auth.html#jupyterhub.auth.Authenticator.authenticate) method.


<details>
<summary>Example</summary>

```
from jupyterhub.auth import Authenticator
import requests

class OPAAuthorizer:
    def __init__(self, opa_url):
        self.opa_url = opa_url

    def is_allowed(self, user, action, resource):
        payload = {
            "input": {
                "user": user,
                "action": action,
                "resource": resource
            }
        }
        response = requests.post(f"{self.opa_url}/v1/data/authz/allow", json=payload)
        
        # If response is successful, return the result (True/False)
        if response.status_code == 200:
            return response.json().get("result", False)
        return False

class OPAAuthenticator(Authenticator):
    def __init__(self, *args, opa_url, **kwargs):
        super().__init__(*args, **kwargs)
        self.authorizer = OPAAuthorizer(opa_url)

    def authenticate(self, handler, data):
        user = data['username']  # Get user info (you can adjust this)
        action = 'login'         # Define action (login for authentication)
        resource = 'jupyterhub'  # Define resource (you can adjust this)
        
        if self.authorizer.is_allowed(user, action, resource):
            # Continue authentication process
            return user
        else:
            # Deny access if not authorized
            raise Exception("Access Denied")

# Use the custom authenticator in JupyterHub config
c.JupyterHub.authenticator_class = OPAAuthenticator
c.OPAAuthenticator.opa_url = "http://<OPA_HOST>:<OPA_PORT>"
```

</details>

### Authorize on spawning a notebook
Provide a custom [spawner class](https://jupyterhub.readthedocs.io/en/stable/reference/api/app.html#jupyterhub.app.JupyterHub.spawner_class) and issue a callout to the OPA server in the start method.

**Comment by adwk67** [comment]
## JupyterHub Operator

A basic Helm install will create several objects.

<details>
<summary>Created on helm install</summary>

```
kind: NetworkPolicy - name: hub
kind: NetworkPolicy - name: proxy
kind: PodDisruptionBudget - name: user-placeholder
kind: ServiceAccount - name: hub
kind: Secret - name: hub
kind: ConfigMap - name: hub
kind: PersistentVolumeClaim - name: hub-db-dir
kind: Role - name: hub
kind: RoleBinding - name: hub
kind: Service - name: hub
kind: Service - name: proxy-api
kind: Service - name: proxy-public
kind: Deployment - name: hub
kind: Deployment - name: proxy
kind: StatefulSet - name: user-placeholder
```

</details>

**Comment by adwk67** [comment]
Waiting on outcome of the decision.

**Comment by lfrancke** [comment]
I have a question on D)

You're only listing advantages of the operator. Are there disadvantages I'm missing?


I'm trying to use my own words, could you tell me if I got it right?

JupyterHub is "only" a central "hub" which people can log in to and then use that JupyterHub UI to "request" actual Notebook instances.
These notebook instances are being "spawned" as new Pods by the "Spawner".

And if we were to create an operator we'd need to decide at which level we want a CRD: At the "hub" or at the "notebook" level.

**Comment by adwk67** [comment]
> JupyterHub is "only" a central "hub" which people can log in to and then use that JupyterHub UI to "request" actual Notebook instances.

Yes, basically. The hub is responsible for authentication and for processing whatever profile has been selected in the JupyterHub Proxy UI.

> These notebook instances are being "spawned" as new Pods by the "Spawner".

Yes, that's my understanding.

> And if we were to create an operator we'd need to decide at which level we want a CRD: At the "hub" or at the "notebook" level.

Yes, and the advantages I listed were only referring to **this** distinction, not the larger question of whether having our own operator makes sense.

**Comment by adwk67** [comment]
> Are there disadvantages I'm missing?

Mainly the effort involved for something that may prove to be much more of a moving target than our other operators. 

- which customers want which stacks?
- how much effort will be involved in addressing CVEs?
- etc.

It **may** be desirable to have an operator that takes a lot away a lot of the configuration pain from the user (for things like Keycloak authentication), but we should be sure the benefits (for us) outweigh the drawbacks.

**Comment by nightkr** [comment]
https://github.com/stackabletech/decisions/issues/36#issuecomment-2681112247

**Comment by adwk67** [comment]
Re-opening as we want to trakc the remaining items of work for the next iteration.

**Comment by adwk67** [comment]
Closing in favour of/superceded by https://github.com/stackabletech/issues/issues/706
